### PR TITLE
Plot model show activations

### LINKED ISSUE #15211: Add layer activations to keras.utils.plot_model and model_to_dot
**System information**.

TensorFlow version (you are using): tf_nightly (keras_nightly-2.7.0.dev2021081907)
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behaviour/state**.

`keras.utils.plot_model` generates a diagram of a model's architecture. This is very useful for debugging and publishing. If layer activations (as opposed to standalone activation layers) are used, for example `tf.keras.layers.Dense(10, activation='relu')`, it does not show the whole architecture because they are not shown. I propose to add an additional parameter: `layer_activations: bool = False` to `keras.utils.plot_model` and `keras.utils.model_to_dot` which when set to True will show the activation for each layer that has one set.

**Will this change the current api? How?**

Yes. It will add an extra parameter, `layer_activations: bool = False`, to both `keras.utils.plot_model` and `keras.utils.model_to_dot`. This is a backwards compatible api change.

It will also potentially formalise more concretely the `activation` attribute as an optional public API of `Layer`. I am not familiar enough with the design of Keras to know the current status of this attribute. There will be no problem for layers that have this attribute or do not have an activation. Layers which store an activation in a different way will be treated incorrectly.

**Who will benefit from this feature?**

Any users who want to generate a complete architecture diagram using `plot_model`, and who specify activation functions by argument, will benefit from this feature.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

I believe the solution is sufficiently specified by the requirements. In case the `layer_activations` argument is False the modified methods should have identical to current behaviour. This will ensure any problems displaying activation functions in unusual cases can be avoided by simply not using this feature.

### PR DESCRIPTION
**Pull request by ddrakard**
This pull request gives library users the option to show layer activations in generated model architecture diagrams.

When the newly added `show_layer_activations` parameter is left at its default value of False, there is no behaviour change (backwards compatible). There is some minor refactoring of existing code to accommodate the changes nicely.

This pull request resolves #15211 .

### CODE REVIEW

**Comment by ddrakard** [comment]
This pull requests implements issue #15211 .

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR. Please add a unit test for this change.

Can you show a screenshot of the new plot output?

**Comment by ddrakard** [comment]
I have added a unit test equivalent to the existing ones: it checks that an output file is created, although it does not verify the contents.

I have attached the diagram generated by the test.

model_5

**Comment by qlzh727** [comment]
@ddrakard, could u point me to the protobuf you want to use? I didn't find anything related to protobuf in this PR.

So far keras only uses protobuf from its own repository, so that we reduce the dependency to tensorflow (eg we don't need to compile any TF protobuf file). The TF PIP package should include the compiled protobuf (py files) in it.

**Comment by ddrakard** [comment]
@qlzh727 Yes, I removed the dependency of protobuf from this PR, but the requirement occurs when I want to use load_model. The `keras/protobuf` directory contains only `.proto` files, not `.py` files. It does not seem that following `CONTRIBUTING.md` causes these to be built. This caused me problems when I wanted to test on my own models.

**Comment by qlzh727** [comment]
The `keras/protobuf` contains the .proto file as expected since we use bazel rule to build the .py file from that. If we keep the compiled py file, we will have to rebuild them everytime we make change to the proto file. On the other hand, the TF pip package contains the compiled py file since the PIP package need to be ready to use without any user compilation. 

If you need to make any change to the TF protobuf file, the change need to made to the TF project first. Keras is a standalone repository and any change that overlap between TF and Keras need to be split into two. 

Could u share more details about load_model? What's the protobuf file you want to use from TF?

**Comment by ddrakard** [comment]
@qlzh727 I have created a new issue #15241 for the protobuf aspect. I hope that is appropriate. It does not obstruct this pull request.

**Comment by ddrakard** [comment]
@qlzh727 I am not clear on the code ownership design of Keras and Tensorflow repositories, given that a lot of the Keras code is duplicated across both. If you could let me know what the contribution process is (does this PR need to go to Tensorflow) I would be grateful.

**Comment by qlzh727** [comment]
> @qlzh727 I am not clear on the code ownership design of Keras and Tensorflow repositories, given that a lot of the Keras code is duplicated across both. If you could let me know what the contribution process is (does this PR need to go to Tensorflow) I would be grateful.

The keras code under tensorflow/python/keras is a duplicate copy and will be removed in the follow up release. We keep it for one release in case ppl are directly importing it somehow, and give them one release cycle to migrate.

I will take a closer look for the issue you filed.

**Comment by innat** [comment]
@ddrakard cool feature. However, what do you think if `plot_model` also provides information about the trainable and non-trainable properties of a layer/layers?

**Comment by ddrakard** [comment]
@innat Thank you.

TL;DR: Needs approval from the Keras team.

I can see uses for that. There are several properties of layers I think would be useful to show. I think a general, optional `hyperparameters_description` (or equivalent) property of `Layer` would be useful, which could for example state the window size and stride of convolution layers. This could optionally be included in the plot. I could add this, but I am not a member of the Keras team and this is a large change that I expect would need approval.

**Comment by fchollet** [comment]
> However, what do you think if plot_model also provides information about the trainable and non-trainable properties of a layer/layers?

Thanks for the suggestion. I would see this as orthogonal to the model's architecture (which is what `plot_model` focuses on). A model's architecture is still the same even if you freeze some layers. So I don't think it's a very useful feature to add to `plot_model`.

**Comment by innat** [comment]
> > However, what do you think if plot_model also provides information about the trainable and non-trainable properties of a layer/layers?
> 
> Thanks for the suggestion. I would see this as orthogonal to the model's architecture (which is what `plot_model` focuses on). A model's architecture is still the same even if you freeze some layers. So I don't think it's a very useful feature to add to `plot_model`.

Yes, I agree with you. 
Generally, I was thinking to have an option that displays or prints some of the most important meta-information or properties of model/layers in a single command (or something similar). But yes, that probably needs a different discussion and is somewhat orthogonal with the `plot_model` function.

**Comment by fchollet** [comment]
This has been merged in https://github.com/keras-team/keras/commit/6caa1073be16075dd41014cc7c7bc2a373dbba01

We had to add new files to the PR hence couldn't merge it directly. Thanks for the great contribution.

**Comment by ddrakard** [comment]
Thank you!
